# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/jsapi/PeerConnectionCtx.cpp
# Commit: fe152aff2a22
# Full Hash: fe152aff2a22e3336dd4696538cbb26779c42fce
# Author: Byron Campen [:bwc] <docfaraday@gmail.com>
# Date: 2022-12-22 17:00:32
# Regressor Bug: 1800136
# File Overlap Count: 1
# Description:
#   Bug 1800136: Fix gap between PC.close() and the storage of its final stats in PeerConnectionCtx. r=ng
#   
#   Ensures that:
#   1) We never have stats report in a still-registered PC _and_ the stats for
#      closed PCs.
# ==============================================================================

diff -r 99ea4f221834 -r fe152aff2a22 dom/media/webrtc/jsapi/PeerConnectionCtx.cpp
--- a/dom/media/webrtc/jsapi/PeerConnectionCtx.cpp	Thu Dec 22 02:33:57 2022 +0000
+++ b/dom/media/webrtc/jsapi/PeerConnectionCtx.cpp	Thu Dec 22 02:33:57 2022 +0000
@@ -32,6 +32,7 @@
 #include "prcvar.h"
 #include "transport/runnable_utils.h"
 #include "WebrtcGlobalChild.h"
+#include "WebrtcGlobalInformation.h"
 
 static const char* pccLogTag = "PeerConnectionCtx";
 #ifdef LOGTAG
@@ -424,9 +425,17 @@
 
 void PeerConnectionCtx::RemovePeerConnection(const std::string& aKey) {
   MOZ_ASSERT(NS_IsMainThread());
-  size_t result = mPeerConnections.erase(aKey);
-  if (mPeerConnections.size() == 0 && result > 0) {
-    mSharedWebrtcState = nullptr;
+  auto it = mPeerConnections.find(aKey);
+  if (it != mPeerConnections.end()) {
+    if (it->second->GetFinalStats() && !it->second->LongTermStatsIsDisabled()) {
+      WebrtcGlobalInformation::StashStats(*(it->second->GetFinalStats()));
+    }
+
+    mPeerConnections.erase(it);
+
+    if (mPeerConnections.empty()) {
+      mSharedWebrtcState = nullptr;
+    }
   }
 }
 
@@ -435,7 +444,7 @@
   MOZ_ASSERT(NS_IsMainThread());
   MOZ_ASSERT(mPeerConnections.count(aKey) == 0,
              "PeerConnection with this key should not already exist");
-  if (mPeerConnections.size() == 0) {
+  if (mPeerConnections.empty()) {
     AudioState::Config audioStateConfig;
     audioStateConfig.audio_mixer = new rtc::RefCountedObject<DummyAudioMixer>();
     AudioProcessingBuilder audio_processing_builder;
@@ -489,6 +498,15 @@
   }
 }
 
+void PeerConnectionCtx::ClearClosedStats() {
+  for (auto& [id, pc] : mPeerConnections) {
+    if (pc->IsClosed()) {
+      // Rare case
+      pc->DisableLongTermStats();
+    }
+  }
+}
+
 nsresult PeerConnectionCtx::Initialize() {
   initGMP();
   SdpRidAttributeList::kMaxRidLength =